### PR TITLE
add os_tag module

### PR DESCRIPTION
**Pull request by gtema**
##### SUMMARY
Add new module `os_tag` to set or remove tags on diverse openstack resources

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME
module.cloud.openstack.os_tag

##### ANSIBLE VERSION

```
ansible 2.7.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/gtema/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 3.7.1 (default, Oct 23 2018, 18:19:07) [GCC 8.2.1 20180801 (Red Hat 8.2.1-2)]

```

##### ADDITIONAL INFORMATION
a new module to set/remove tags on openstack resources (which support that) is added.

### CODE REVIEW

**Comment by mnaser** [comment]
@gtema any reason you're not just extending `os_server`?

**Comment by gtema** [comment]
@mnaser, not really, but:
- tags are extensions
- tags are on the "separate group" of API description
- there are server_action, server_group, server_metadata, etc, which are separated, so I followed this pattern not to complicate os_server module
- changing tags does not require (and does not support) setting being part of the "server" request, so it does not really fit os_server.present state (can be of course discussed, but otherwise would require merge of actions, metadata, group etc.)

So basically just KISS and following present pattern.

**Comment by mnaser** [comment]
That makes a lot of sense!

**Comment by mnaser** [comment]
ship

**Comment by mnaser** [comment]
shipit

File: lib/ansible/modules/cloud/openstack/os_server_tag.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+                    'supported_by': 'community'}
+
+
+DOCUMENTATION = '''
+---
+module: os_server_tag
+short_description: Manage server tags
+extends_documentation_fragment: openstack
+version_added: "2.8"
+author: "Artem Goncharov (gtema)"
```
Comment (resmo) [inline lib/ansible/modules/cloud/openstack/os_server_tag.py]:
~~~diff
- author: "Artem Goncharov (gtema)"
+ author: Artem Goncharov (@gtema)

File: lib/ansible/modules/cloud/openstack/os_server_tag.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    - set or delete tag on the Nova server
+options:
+   server:
+      description:
+        - Name or id of the server
+      required: true
+   state:
+     description:
+       - Should the resource be present or absent.
+     choices: [present, absent]
```
Comment (resmo) [inline lib/ansible/modules/cloud/openstack/os_server_tag.py]:
~~~diff
-     choices: [present, absent]
+     choices: [ present, absent ]
~~~

File: lib/ansible/modules/cloud/openstack/os_server_tag.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    state: present
+    name: __crazy_tag
+'''
+
+RETURN = '''
+server:
+    description: UUID of the server.
+    returned: success
+    type: string
+changed:
```
Comment (resmo) [inline lib/ansible/modules/cloud/openstack/os_server_tag.py]: this is a common return and must not be documented in a module, please remove it.

File: lib/ansible/modules/cloud/openstack/os_server_tag.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+     description:
+       - Ignored. Present for backwards compatibility
+requirements:
+    - "python >= 2.7"
+    - "openstacksdk"
+'''
+EXAMPLES = '''
+---
+- name: set the tag on the instance
+  ignore_errors: True
```
Comment (resmo) [inline lib/ansible/modules/cloud/openstack/os_server_tag.py]: why is ignore_errors  needed?

File: lib/ansible/modules/cloud/openstack/os_server_tag.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+    argument_spec = openstack_full_argument_spec(
+        server=dict(required=True),
+        state=dict(default='present', choices=['absent', 'present']),
+        tag=dict(required=True),
+    )
+
+    module_kwargs = openstack_module_kwargs()
+    module = AnsibleModule(argument_spec,
+                           supports_check_mode=True,
```
Comment (resmo) [inline lib/ansible/modules/cloud/openstack/os_server_tag.py]: I don't see any check mode support implemented in the code. Is check mode really supported?

**Review by resmo — CHANGES_REQUESTED** [review]

File: changelogs/fragments/add-os-server-tag.yaml [hunk 6]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
```
Comment (resmo) [inline changelogs/fragments/add-os-server-tag.yaml]: please remove this changelog fragment, this is going to be auto generated for new modues.

File: lib/ansible/modules/cloud/openstack/os_server_tag.py (hunk 1 above)
Comment (gtema) [reply lib/ansible/modules/cloud/openstack/os_server_tag.py]: Done

File: lib/ansible/modules/cloud/openstack/os_server_tag.py (hunk 2 above)
Comment (gtema) [reply lib/ansible/modules/cloud/openstack/os_server_tag.py]: Done

File: lib/ansible/modules/cloud/openstack/os_server_tag.py (hunk 3 above)
Comment (gtema) [reply lib/ansible/modules/cloud/openstack/os_server_tag.py]: Done

File: lib/ansible/modules/cloud/openstack/os_server_tag.py (hunk 4 above)
Comment (gtema) [reply lib/ansible/modules/cloud/openstack/os_server_tag.py]: not needed. A copy artifact. Removed

File: lib/ansible/modules/cloud/openstack/os_server_tag.py (hunk 5 above)
Comment (gtema) [reply lib/ansible/modules/cloud/openstack/os_server_tag.py]: Nope, removed

File: changelogs/fragments/add-os-server-tag.yaml (hunk 6 above)
Comment (gtema) [reply changelogs/fragments/add-os-server-tag.yaml]: Done

**Review by resmo — CHANGES_REQUESTED** [review]
I am missing idempotency, currently the module does not check if the tag exists and always show "changed", right? It should only put ar delete a tag if needed to get to the state declared.
This would also allow to implement check_mode support.
Last, it does not make sense to return changed=false if the server was not found. I would expect the module would fail with "could not find server xy".

**Comment by gtema** [comment]
ready_for_review

**Comment by sshnaidm** [comment]
Thanks for submitting patch for Openstack Ansible modules!
We moved Openstack Ansible modules to Openstack repositories.
Next patches should be submitted not with Ansible Github but with
Openstack Gerrit: https://review.opendev.org/#/q/project:openstack/ansible-collections-openstack
Please submit your code there from now.
Thanks for your contribution and sorry for inconvienience.

File: lib/ansible/modules/cloud/openstack/os_tag.py [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+            url_prefix = '/ports'
+            endpoint = cloud.network
+            resource = port
+        elif router:
+            instance = cloud.get_router(router)
+            url_prefix = '/routers'
+            endpoint = cloud.network
+            resource = router
+        elif security_group_rule:
+            instance = cloud.get_security_group_rule(security_group_rule)
```
Comment (MaysaMacedo) [inline lib/ansible/modules/cloud/openstack/os_tag.py:296]: The security group rule and the trunk are not present in any of the resources' contents here: https://github.com/openstack/openstacksdk/tree/master/openstack/cloud, consequently using the `cloud.get_<resource>` call fails. Instead the `could.network` call can be used.

File: lib/ansible/modules/cloud/openstack/os_tag.py [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+extends_documentation_fragment: openstack
+version_added: "2.8"
+author: Artem Goncharov (@gtema)
+description:
+    - set or delete tags on the OpenStack resources
+options:
+   server:
+      description:
+        - Name or id of the Nova Server resource.
+   floating_ip:
```
Comment (MaysaMacedo) [inline lib/ansible/modules/cloud/openstack/os_tag.py:30]: This documentation may need update, as according to the code[1] Floating IP can only be filtered by ID.

[1 ]https://github.com/openstack/openstacksdk/blob/master/openstack/cloud/_floating_ip.py#L96-L99

**Review by MaysaMacedo — COMMENTED** [review]
With the migration, do you plan to move the work proposed here to `openstack/ansible-collections-openstack` repository on gerrit?

**Comment by gtema** [comment]
Yes, it would be moved

**Pull request closed without merging**